### PR TITLE
[REPL] Fix caching of unresolved imports

### PR DESCRIPTION
**Pull request by ileasile**
Before this fix, if some imports were not resolved during compilation, this result had been saved in caches, and this import couldn't been resolved during following compilations even if it was added to the module dependencies. This PR adds special handling of packages resolution caches for the REPL compiler.

Also this PR fixes performance problem with updating of configuration for implicit receivers.

### CODE REVIEW

File: compiler/resolution.common.jvm/src/org/jetbrains/kotlin/resolve/jvm/KotlinJavaPsiFacade.java [hunk 1]
```diff
@@ -353,24 +360,22 @@ public PsiPackage findPackage(@NotNull String qualifiedName, GlobalSearchScope s
...
-                cache.packageInLibScopeCache : cache.packageInScopeCache;
+        ConcurrentMap<Pair<String, GlobalSearchScope>, PsiPackage> notFoundPackageInScopeCache;
+        switch (notFoundCacheType) {
+            case LIB_SCOPE:
+                notFoundPackageInScopeCache = cache.packageInLibScopeCache;
+                break;
+            case SCOPE:
+                notFoundPackageInScopeCache = cache.packageInScopeCache;
+                break;
+            default:
```
Comment (dzharkov) [inline compiler/resolution.common.jvm/src/org/jetbrains/kotlin/resolve/jvm/KotlinJavaPsiFacade.java:376]: I'd explicitly enumerated NO_CACHING here

File: core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/PackageViewDescriptorAccessor.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.jetbrains.kotlin.descriptors.impl
+
+import org.jetbrains.kotlin.descriptors.ModuleCapability
+import org.jetbrains.kotlin.descriptors.PackageViewDescriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.storage.StorageManager
+
+interface PackageViewDescriptorAccessor {
```
Comment (dzharkov) [inline core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/PackageViewDescriptorAccessor.kt:13]: The name like `PackageViewDescriptor` seems to be more clear to me, but I don't insist here

**Review by ligee — APPROVED** [review]

File: compiler/resolution.common.jvm/src/org/jetbrains/kotlin/resolve/jvm/KotlinJavaPsiFacade.java (hunk 1 above)
Comment (ileasile) [reply compiler/resolution.common.jvm/src/org/jetbrains/kotlin/resolve/jvm/KotlinJavaPsiFacade.java]: Handled this case

**Comment by ileasile** [comment]
This branch was merged manually

**Pull request closed without merging**